e2e/stack_test.go

```diff
@@ -282,48 +272,244 @@ func depositE2E(t *testing.T, stack *e2e.StackConfig) {
 	)
+	result := requireL2Event(t, stack, query)
+
+	require.NotEmpty(t, result.Txs, "mint_eth event not found")
+	t.Log("Monomer can mint ETH for L1 user deposits")
+}
+
+func requireEthIsBurnedAndGetL2BlockNum(t *testing.T, stack *e2e.StackConfig, userAddress, valueHex string) *big.Int {
+	query := fmt.Sprintf(
+		"%s.%s='%s' AND %s.%s='%s' AND %s.%s='%s'",
+		rolluptypes.EventTypeBurnETH, rolluptypes.AttributeKeyL2WithdrawalTx, rolluptypes.EventTypeWithdrawalInitiated,
+		rolluptypes.EventTypeBurnETH, rolluptypes.AttributeKeyFromCosmosAddress, userAddress,
+		rolluptypes.EventTypeBurnETH, rolluptypes.AttributeKeyValue, valueHex,
+	)
+	result := requireL2Event(t, stack, query)
+
+	require.NotEmpty(t, result.Txs, "burn_eth event not found")
+	t.Log("Monomer can burn ETH for L2 user withdrawals")
+
+	return big.NewInt(result.Txs[0].Height)
+}
+
+func requireL2Event(t *testing.T, stack *e2e.StackConfig, query string) *cometcore.ResultTxSearch {
 	page := 1
 	perPage := 100
-	orderBy := "desc"
 
-	result, err := appchainClient.TxSearch(
-		context.Background(),
+	result, err := stack.L2Client.TxSearch(
+		stack.Ctx,
 		query,
 		false,
 		&page,
 		&perPage,
-		orderBy,
+		"desc",
 	)
 	require.NoError(t, err, "search transactions")
 	require.NotNil(t, result)
-	require.NotEmpty(t, result.Txs, "mint_eth event not found")
-	t.Log("Monomer can mint_eth from L1 user deposits")
+	return result
+}
+
+func createL1TransactOpts(
+	t *testing.T,
+	stack *e2e.StackConfig,
+	user *e2e.L1User,
+	l1signer types.Signer,
+	l1GasLimit uint64,
+	value *big.Int,
+) *bind.TransactOpts {
+	return &bind.TransactOpts{
+		From: user.Address,
+		Signer: func(addr common.Address, tx *types.Transaction) (*types.Transaction, error) {
+			signed, err := types.SignTx(tx, l1signer, user.PrivateKey)
+			if err != nil {
+				return nil, err
+			}
+			return signed, nil
+		},
+		Nonce:    getCurrentUserNonce(t, stack, user.Address),
+		GasPrice: getSuggestedL1GasPrice(t, stack),
+		GasLimit: l1GasLimit,
+		Value:    value,
+		Context:  stack.Ctx,
+		NoSend:   false,
+	}
+}
+
+func getCurrentUserNonce(t *testing.T, stack *e2e.StackConfig, userAddress common.Address) *big.Int {
+	nonce, err := stack.L1Client.NonceAt(stack.Ctx, userAddress, nil)
+	require.NoError(t, err)
+	return new(big.Int).SetUint64(nonce)
+}
+
+func getSuggestedL1GasPrice(t *testing.T, stack *e2e.StackConfig) *big.Int {
+	gasPrice, err := stack.L1Client.Client.SuggestGasPrice(stack.Ctx)
+	require.NoError(t, err)
+	return gasPrice
+}
+
+func createL1FilterOpts(ctx context.Context) *bind.FilterOpts {
+	return &bind.FilterOpts{
+		Start:   0,
+		End:     nil,
+		Context: ctx,
+	}
+}
+
+// waitForL2OutputProposal waits for the L2 output containing the withdrawal tx to be proposed on L1 and returns
+// the block number of the L2 output proposal.
+func waitForL2OutputProposal(t *testing.T, l2OutputOracleCaller *bindings.L2OutputOracleCaller, l2BlockNumber *big.Int) *big.Int {
```

Comment:
Question for reviewers - Instead of waiting for the output block to be proposed, we could alternatively just call the `nextBlockNumber` func in the `L2OutputOracle` contract and then wait until that L2 block. Lmk which approach you think is cleaner

Reply:
I convinced myself that just calling `nextBlockNumber` would be a lot better, I'll add a new commit with these changes

Reply:
Unfortunately I couldn't just use `stack.WaitL1/L2` since the op-proposer has some variance between tests. I still updated this helper func to make it clearer and hopefully more readable in https://github.com/polymerdao/monomer/pull/206/commits/455ac42cc7cf6ac1435f56c19acfb1b52ed528db